sources/packages/backend/apps/api/src/route-controllers/education-program-offering/_tests_/e2e/education-program-offering.institutions.controller.bulkInsert.e2e-spec.ts

```diff
@@ -0,0 +1,441 @@
+  User,
+} from "@sims/sims-db";
+import {
+  E2EDataSources,
+  createE2EDataSources,
+  createFakeInstitutionLocation,
+} from "@sims/test-utils";
+import {
+  createTestingAppModule,
+  BEARER_AUTH_TYPE,
+  InstitutionTokenTypes,
+  getInstitutionToken,
+  authorizeUserTokenForLocation,
+  getAuthRelatedEntities,
+  createFakeEducationProgram,
+} from "../../../../testHelpers";
+import * as request from "supertest";
+import * as path from "path";
+import { In } from "typeorm";
+import { OFFERING_VALIDATION_CRITICAL_ERROR } from "../../../../constants";
+import { OfferingValidationWarnings } from "../../../../services";
+
+describe("EducationProgramOfferingInstitutionsController(e2e)-bulkInsert", () => {
+  let app: INestApplication;
+  let db: E2EDataSources;
+  let collegeF: Institution;
+  let collegeFUser: User;
+  let institutionUserToken: string;
+  let endpoint: string;
+  let csvLocationCodeYESK: string;
+  let csvProgramSABCCodeSBC2: string;
+  let csvLocationCodeKSEY: string;
+  let csvProgramSABCCodeSBC4: string;
+  let multipleOfferingFilePath: string;
+  let singleOfferingFilePath: string;
+  let singleOfferingWithValidationErrorsFilePath: string;
+
+  beforeAll(async () => {
+    const { nestApplication, dataSource } = await createTestingAppModule();
+    app = nestApplication;
+    db = createE2EDataSources(dataSource);
+    const { institution, user: institutionUser } = await getAuthRelatedEntities(
+      db.dataSource,
+      InstitutionTokenTypes.CollegeFUser,
+    );
+    collegeF = institution;
+    collegeFUser = institutionUser;
+
+    // location code in the single and multiple CSV.
+    csvLocationCodeYESK = "YESK";
+    // SABC code in the single and multiple CSV.
+    csvProgramSABCCodeSBC2 = "SBC2";
+    // Second location code in the multiple CSV.
+    csvLocationCodeKSEY = "KSEY";
+    // Second SABC code in the multiple CSV.
+    csvProgramSABCCodeSBC4 = "SBC4";
+
+    institutionUserToken = await getInstitutionToken(
+      InstitutionTokenTypes.CollegeFUser,
+    );
+    endpoint = "/institutions/education-program-offering/bulk-insert";
+
+    multipleOfferingFilePath = path.join(
+      __dirname,
+      "bulk-insert/multiple-upload.csv",
+    );
+    singleOfferingFilePath = path.join(
+      __dirname,
+      "bulk-insert/single-upload.csv",
+    );
+    singleOfferingWithValidationErrorsFilePath = path.join(
+      __dirname,
+      "bulk-insert/single-upload-with-validation-errors.csv",
+    );
+  });
+
+  it(
+    "Should create an approved and a creation pending offerings when a multi line bulk" +
+      " offering CSV file with existing location code and SABC code, and one with" +
+      " same delivery method and another with different delivery method is uploaded.",
+    async () => {
+      // Arrange
+      // Creating an institution location with same location code as that of the
+      // first row of the multiple CSV file.
+      const collegeFLocationYESK = createFakeInstitutionLocation(
```

Comment:
The API test suits run in parallel, hence there is no guarantee that the location with institution code YESK (in this case) could not be created before. 

I had the same situation while creating e2e tests for ece response integration. I followed this approach and it worked.
![image](https://github.com/bcgov/SIMS/assets/54600590/d71629da-0966-4ff8-966b-652c7c7883d4)

We could possibility centralize and re-use the same method. Let me know if I am missing something.